tools: add release tool, remove old tools (WIP)

rvagg opened the pull request:
Not finished, pushing now in case people want to give feedback.

This is a release tool to be used by those authorized by the TC to create and promote releases. The person doing the releases will need to:

 * Have a commit-bit to be able to commit in src/node_version.h, update the changelog and tag a release
 * Have access to Jenkins to run the *iojs+release* job
 * Be able to access the dist@iojs.org account by having their SSH key on the server (the _dist_ user controls the /releases/ directory, the _staging_ user is used by the release build servers to upload their binaries prior to them being promoted by the _dist_ account).
 * Have their GPG key fingerprint listed on the README so the SHASUMS file for their release(s) can be verified

This is nearly ready, I just have some minor items to sort out but we'll get this in to use for the 1.1.0 in #674 where @chrisdickinson will be my guinea-pig! There should also be a document in doc/ to go along with this to explain the process and how to cut a release.

I also need to update the server-side scripts that pair with this in the https://github.com/iojs/build repo, some of it is already there but I need to be able to publicly show the whole lot of this process.

rvagg commented:
Should be ready to merge now I think. Please review @chrisdickinson (and anyone else that wants to).

I've added a verbose releases.md document that describes the release process for anyone authorized to make a release. I don't see this as a final state but simply documenting how it works now; I imagine we'll be iterating on this over time and we'll need to keep this updated.

Also, @chrisdickinson, can you give me your SSH pubkey, you have 3 listed [here](https://github.com/chrisdickinson.keys).

I just need to get Chris' key onto the web server and this is ready to roll I think. I also need to get more stuff into the build repo so I can show what's actually running on the server to make this work, that's not blocking a release but I'll get to it ASAP.

rvagg commented:
Oh, I changed the README to include a section on downloading binaries and some stuff about how to verify checksums and verify checksum files using GPG.

jbergstroem commented on tools/release.sh:
```diff
@@ -0,0 +1,121 @@
+#!/bin/bash
```
how about env bash instead? bash lives elsewhere on FreeBSD and some other systems.

jbergstroem commented on tools/release.sh:
```diff
@@ -0,0 +1,121 @@
+    gpg --default-key $gpgkey --clearsign ${tmpdir}/${shafile}
+    gpg --default-key $gpgkey --armor --export --output ${tmpdir}/${shafile}.gpg
+
+    scp ${tmpdir}/${shafile} ${tmpdir}/${shafile}.asc ${tmpdir}/${shafile}.gpg ${webuser}@${webhost}:${shadir}/
```
There might be a point in separating the actual upload so the person doing the release can review it "one last time" before a potential destructive action.

rvagg commented:
* changed to `#!/usr/bin/env bash` as per @jbergstroem's suggestion
* introduced a `cat` and prompt once SHASUMS256.txt has been signed to allow the user to see what they are uploading, as per @jbergstroem's suggestion
* introduced a check that the git tag for `vx.y.z` is signed with the same key you're using to sign a release
* introduced a separate `release.sh -s version` mode where you can do _only_ the checksum & sign step for a version that's already been prompted--in case something goes wrong with the original signing (like you get your GPG password wrong 3 times and it bails)

Fishrock123 commented on tools/build-changelog.sh:
```diff
@@ -1,16 +0,0 @@
-  | sed -E 's|__VERSION__|v'$(python tools/getnodeversion.py)'|g' \
-  > out/doc/changelog.html
-
-rm out/doc/changelog-body.html
```
What do we use for the changelog now?

rvagg replied:
nada, that stuff was for changelog.html, which we don't publish (yet); we just point people to CHANGELOG.md currently.

When we come up with a better strategy we can fix this, however I think this might be something that the @iojs/website team takes responsibility for building something for coming up wtih something much prettier for.

For details on how to build the CHANGELOG.md entry itself, see doc/releases.md above.

rvagg commented:
* Removed the stale changelog & email targets from the Makefile, we don't use them and if we do I'm sure the website team can come up with something better

bnoordhuis commented on README.md:
```diff
@@ -209,6 +265,7 @@ information about the governance of the io.js project, see
 * **Colin Ihrig** ([@cjihrig](https://github.com/cjihrig)) &lt;cjihrig@gmail.com&gt; (Technical Committee)
 * **Mikeal Rogers** ([@mikeal](https://github.com/mikeal)) &lt;mikeal.rogers@gmail.com&gt;
 * **Rod Vagg** ([@rvagg](https://github.com/rvagg)) &lt;rod@vagg.org&gt;
+<br>Release GPG key: DD8F2338BAE7501E3DD5AC78C273792F7D83545D
```
Maybe a naive question but can't we use a single master key with as many sub-keys as there are release managers?  It would make all releases be signed by "io.js Release Engineering &lt;releng@iojs.org&gt;" (or whatever we decide to call ourselves) and it makes revocation a little easier.  Master key management is an open issue, though; who gets to own it?

rvagg replied:
Good question, to which I don't have a simple answer unfortunately.

My personal bias is towards decentralising this and making individuals responsible for signing their releases with their own keys, simply because I think decentralisation in most aspects of the project are more healthy for the community focus and ethos that we're trying to foster. It also makes that individual properly responsible for that release rather than it just being a mechanical task.

However, I don't have any strong technical objections to a more rigid signing mechanism if others prefer that. Perhaps a question for @iojs/tc.

bnoordhuis commented on doc/releases.md:
````diff
@@ -0,0 +1,200 @@
+
+```
+#define NODE_VERSION_IS_RELEASE 1
+```
````
This section should explain when to bump NODE_MODULE_VERSION and what it means for the version number (at least a minor bump; I don't think we've decided on a major bump policy yet.)

rvagg replied:
good point, adding this:

**Also consider whether to bump `NODE_MODULE_VERSION`**:

This macro is used to signal an ABI version for native addons. It currently has two common uses in the community:

* Determining what API to work against for compiling native addons, e.g. [NAN](https://github.com/rvagg/nan) uses it to form a compatibility-layer for much of what it wraps.
* Determining the ABI for downloading pre-built binaries of native addons, e.g. [node-pre-gyp](https://github.com/mapbox/node-pre-gyp) uses this value as exposed via `process.versions.modules`
to help determine the appropriate binary to download at install-time.

The general rule is to bump this version when there are _breaking ABI_ changes and also if there are non-trivial API changes. The rules are not yet strictly defined, so if in doubt, please conf
er with someone that will have a more informed perspective, such as a member of the NAN team.

chrisdickinson commented on doc/releases.md:
```diff
@@ -0,0 +1,209 @@
+
+A SHASUMS256.txt file is produced for every promoted build, nightly and releases. Additionally for releases, this file is signed by the individual responsible for that release. In order to be able to verify downloaded binaries, the public should be able to check that the SHASUMS256.txt file has been signed by someone who has been authorized to create a release.
+
+The GPG keys should be fetchable from a known third-party keyserver, currently the SKS Keyservers at <https://sks-keyservers.net> are recommended. Use the [submission](https://sks-keyservers.net/i/#submit) form to submit a new GPG key. Keys should be fetchable via:
```
I'm getting an insecure connection warning when I try to submit my pubkey to this server. Is this something I should be worried about?

indutny replied:
Btw, is there any reason to not use https://pgp.mit.edu/ ?

chrisdickinson replied:
Eventually went ahead with sending my pubkey despite the insecure connection. Was rewarded with a 500 Internal Server Error. ):

rvagg replied:
The only reason I chose sks-keyservers.net was because the only precedent I'm aware of for checking signed shasums is in the Docker images for both Node (official Docker ones) and io.js and that's what they've chosen to use. I don't know the reasoning behind this but I went with that as an established convention.

I'm happy to change based on recommendation, @bnoordhuis also had some thoughts about PGP keys so we should chat about this tomorrow at TC.

indutny replied:
Ok, sounds like a good topic for @iojs/tc agenda. @rvagg let's land it the way it is for now, we will fix it later.

bnoordhuis commented:
LGTM FWIW

rvagg commented:
landed @ df48faf, thanks

chrisdickinson commented on doc/releases.md:
````diff
@@ -0,0 +1,209 @@
+
+```
+git push origin branch vx.y.z
+# where "branch" is the working branch and "vx.y.z" the the release version
````
(we may later add `e.g., v1.x` here)

chrisdickinson commented on doc/releases.md:
````diff
@@ -0,0 +1,209 @@
+gpg --keyserver pool.sks-keyservers.net --recv-keys <FINGERPRINT>
+```
+
+Additionally, full GPG key fingerprints for individuals authorized to release should be listed in the io.js GitHub README.md file.
````
Yep, I totally missed this :|
